src/main/app/src/app/admin/parameter-edit/smart-documents-form/smart-documents-form.component.ts

```diff
@@ -0,0 +1,346 @@
+import { FormGroup } from "@angular/forms";
+import {
+  MatTreeFlatDataSource,
+  MatTreeFlattener,
+} from "@angular/material/tree";
+import { injectQuery } from "@tanstack/angular-query-experimental";
+import { firstValueFrom, Observable } from "rxjs";
+import { InformatieObjectenService } from "src/app/informatie-objecten/informatie-objecten.service";
+import { GeneratedType } from "src/app/shared/utils/generated-types";
+import { SmartDocumentsService } from "../../smart-documents.service";
+
+type SmartDocumentsTemplateWithParentId =
+  GeneratedType<"RestSmartDocumentsTemplate"> & {
+    parentGroupId: string;
+  };
+type SmartDocumentsTemplateGroupWithParentId = Omit<
+  GeneratedType<"RestSmartDocumentsTemplateGroup">,
+  "groups" | "templates"
+> & {
+  groups?: SmartDocumentsTemplateGroupWithParentId[] | null;
+  templates?: SmartDocumentsTemplateWithParentId[] | null;
+};
+
+type MappedSmartDocumentsTemplateWithParentId =
+  GeneratedType<"RestMappedSmartDocumentsTemplate"> & {
+    parentGroupId?: string | null;
+  };
+type MappedSmartDocumentsTemplateGroupWithParentId = Omit<
+  GeneratedType<"RestMappedSmartDocumentsTemplateGroup">,
+  "groups" | "templates"
+> & {
+  groups?: MappedSmartDocumentsTemplateGroupWithParentId[] | null;
+  templates?: MappedSmartDocumentsTemplateWithParentId[] | null;
+};
+
+type PlainTemplateMappings = Omit<
+  MappedSmartDocumentsTemplateWithParentId,
+  "name"
+>;
+
+type MappedSmartDocumentsTemplateFlattenedGroupWithParentId = Omit<
+  MappedSmartDocumentsTemplateGroupWithParentId,
+  "groups"
+>;
+
+interface FlatNode {
+  expandable: boolean;
+  name: string;
+  level: number;
+}
+
+@Component({
+  selector: "smart-documents-form",
+  templateUrl: "./smart-documents-form.component.html",
+  styleUrl: "./smart-documents-form.component.less",
+})
+export class SmartDocumentsFormComponent {
+  @Input() formGroup: FormGroup;
+  @Input() zaakTypeUuid: string;
+  @Output() formValidityChanged = new EventEmitter<boolean>();
+
+  allSmartDocumentTemplateGroups: GeneratedType<"RestSmartDocumentsTemplateGroup">[] =
+    [];
+  currentTemplateMappings: MappedSmartDocumentsTemplateGroupWithParentId[] = [];
+  informationObjectTypes: GeneratedType<"RestInformatieobjecttype">[] = [];
+
+  newTemplateMappings: MappedSmartDocumentsTemplateGroupWithParentId[] = [];
+
+  constructor(
+    private smartDocumentsService: SmartDocumentsService,
+    private informatieObjectenService: InformatieObjectenService,
+  ) {
+    effect(() => this.prepareDatasource());
+  }
+
+  ngOnInit() {
+    this.formGroup.statusChanges.subscribe(() => {
+      this.formValidityChanged.emit(this.formGroup.valid);
+    });
+  }
+
+  private prepareDatasource() {
+    this.allSmartDocumentTemplateGroups =
+      this.allSmartDocumentTemplateGroupsQuery.data()
+        ? this.addParentIdToTemplates(
+            this.allSmartDocumentTemplateGroupsQuery.data(),
+          )
+        : [];
```

Comment:
This could be made a bit simpler

either make sure `addParentIdToTemplates` can handle `undefined` by always returning `[]` if passed `undefined` or always pass `[]` like"

```
this.addParentIdToTemplates(
  this.allSmartDocumentTemplateGroupsQuery.data() ?? [],
);
```

Reply:
Fixed ✅